src/charm.py

```diff
@@ -277,6 +278,146 @@ def is_unit_busy(self) -> bool:
+        )
+
+    def _join_unit_to_cluster(self) -> None:
+        """Join the unit to the cluster.
+
+        Try to join the unit from the primary unit.
+        """
+        instance_label = self.unit.name.replace("/", "-")
+        instance_fqdn = self._get_unit_fqdn(self.unit.name)
+
+        if self._mysql.is_instance_in_cluster(instance_label):
+            logger.debug("instance already in cluster")
+            return
+
+        # Add new instance to the cluster
+        try:
+            cluster_primary = self._get_primary_address_from_peers()
+            if not cluster_primary:
+                self.unit.status = WaitingStatus("waiting to get cluster primary from peers")
+                logger.debug("waiting: unable to retrieve the cluster primary from peers")
+                return
+
+            if self._mysql.get_cluster_node_count(from_instance=cluster_primary) == GR_MAX_MEMBERS:
+                self.unit.status = BlockedStatus(
+                    f"Cluster reached max size of {GR_MAX_MEMBERS} units. Standby."
+                )
+                logger.warning(
+                    f"Cluster reached max size of {GR_MAX_MEMBERS} unit. This unit will stay as standby."
+                )
+                return
+
+            self.unit.status = MaintenanceStatus("joining the cluster")
+
+            self._mysql.add_instance_to_cluster(
+                instance_fqdn, instance_label, from_instance=cluster_primary
+            )
+            logger.debug(f"Added instance {instance_fqdn} to cluster")
+
+            # Update 'units-added-to-cluster' counter in the peer relation databag
+            self.unit_peer_data["unit-initialized"] = "True"
+            self.unit_peer_data["member-state"] = "online"
+            self.unit.status = ActiveStatus(self.active_status_message)
+            logger.debug(f"Instance {instance_label} is cluster member")
+
+        except MySQLAddInstanceToClusterError:
+            logger.debug(f"Unable to add instance {instance_fqdn} to cluster.")
+        except MySQLLockAcquisitionError:
+            self.unit.status = WaitingStatus("waiting to join the cluster")
+            logger.debug("Waiting to joing the cluster, failed to acquire lock.")
+
+    def _remove_scaled_down_units(self) -> None:
+        """Remove scaled down units from the cluster."""
+        planned_units = self.app.planned_units()
+        cluster_status = self._mysql.get_cluster_status()
+        if not cluster_status:
+            self.unit.status = BlockedStatus("Failed to get cluster status")
+            return
+
+        try:
+            addresses_of_units_to_remove = [
+                member["address"]
+                for unit_name, member in cluster_status["defaultreplicaset"]["topology"].items()
+                if int(unit_name.split("-")[-1]) >= planned_units
+            ]
+        except ValueError:
+            # exception can occur if unit is not yet labeled
```

Comment:
if labeling not in place, the `split("-")` will fail because unit name is not like `mysql-0`.
tbf, this will (hopefully) soon be refactored if scaledown on juju works as expected